### DISCUSSION TITLE
"Neutral" tonemapping darkens the image considerably

### DISCUSSION THREAD

**Discussion by hybridherbst** [google/model-viewer#4830] [Q&A]
### Description

For one of our viewers, I was using an earlier version of Neutral / Khronos PBR Tonemapper as it was in use by model-viewer. In that version, there was no/almost no darkening on surfaces compared to Linear/None, and only a curve in the highlights to prevent clipping, to the best of my understanding. 

To my surprise, switching to the current version of "Neutral" now darkens the image considerably – which doesn't feel intended, and is different to before. Darker areas are now so dark that it loses detail entirely compared to the Linear look (see for example the exhaust area in the pictures below, or the rubber pipes in the back).

I see the same behavior in both three.js and model-viewer, so I think raising the issue here as the "source" of the tonemapper is appropriate.

Long story short – is this intended? Does the shipped version of the Khronos PBR Tonemapper now resemble ACES so much more than previous in-progress versions? To me, that feels like a step back – previously my understanding was that Neutral is designed to keep albedo data as much as possible, but now it feels the goal is to be as close as possible to ACES without having the same hue shifts?

#### Live Demo – Damaged Helmet

- Open https://modelviewer.dev/editor
- Load DamagedHelmet ([DamagedHelmet (3).glb.zip](https://github.com/user-attachments/files/16014044/DamagedHelmet.3.glb.zip))
- Load Venice Sunrise for comparison ([venice_sunrise_1k.hdr.zip](https://github.com/user-attachments/files/16014266/venice_sunrise_1k.hdr.zip))
- Open https://threejs.org/examples/webgl_tonemapping.html in parallel

Alternative reproduction: Neutral also darkens "Damaged Helmet" considerably on https://tone-mapping.glitch.me/.

| Tone Mapping | model-viewer | three.js |
| - | - | - |
| Linear | <img width="626" alt="Screenshot 2024-06-27 at 14 42 01" src="https://github.com/google/model-viewer/assets/2693840/c1306d5c-bbe1-4346-a38a-0559966deca2"> | <img width="694" alt="Screenshot 2024-06-27 at 14 44 42" src="https://github.com/google/model-viewer/assets/2693840/2b2798e0-887d-4cfc-aeb8-f6be5fdfaf9e"> |
| Neutral | <img width="624" alt="Screenshot 2024-06-27 at 14 42 14" src="https://github.com/google/model-viewer/assets/2693840/2eb17552-121f-41d8-9edd-01029faa32a8"> | <img width="698" alt="Screenshot 2024-06-27 at 14 44 52" src="https://github.com/google/model-viewer/assets/2693840/0585e673-799e-4ac8-bab6-26a52527d1b0"> |
| ACES | <img width="625" alt="Screenshot 2024-06-27 at 14 42 22" src="https://github.com/google/model-viewer/assets/2693840/a0ed59f0-4d4f-4efb-a512-407c38bf7c2e">  | <img width="680" alt="Screenshot 2024-06-27 at 14 45 02" src="https://github.com/google/model-viewer/assets/2693840/55b0f264-1515-45f4-8fad-44f129848dce"> |
| Base Color Map (Reference) | ![image](https://github.com/google/model-viewer/assets/2693840/19cac65c-b159-4b72-a796-bde21dd48284) | ![image](https://github.com/google/model-viewer/assets/2693840/19cac65c-b159-4b72-a796-bde21dd48284) |

Since the model-viewer editor doesn't support switching to None/Linear for comparison (#4541), you can paste and run this code snippet in the chrome console to switch to Linear:

```js
const mv = document.querySelector("model-viewer-preview").shadowRoot.querySelector("model-viewer");
const sc = mv[Object.getOwnPropertySymbols(mv).find((x) => x.description === "scene")];
sc.toneMapping = 1;
sc.queueRender();
```

#### Live Demo – Caterpillar Work Boot

Here's another real-world scanned PBR model that fares even worse with Neutral tonemapping – I'd say Neutral looks the worst of all tonemapping options, heavily deviating from the intended look of the model:

| Tone Mapping | Rendering |
| - | - |
| Base Color (Map * Factor) Reference | ![image](https://github.com/google/model-viewer/assets/2693840/a0ba220d-3b93-471e-963b-be1aab94ddf0) |
| Linear |<img width="802" alt="Screenshot 2024-06-27 at 15 47 27" src="https://github.com/google/model-viewer/assets/2693840/810ff1e3-b8ea-41d9-8f45-903336e0bbe6"> | 
| Neutral | <img width="807" alt="Screenshot 2024-06-27 at 15 47 53" src="https://github.com/google/model-viewer/assets/2693840/1cd72888-2df3-431c-9198-4eee2ac579d3"> | 
| AgX | <img width="803" alt="Screenshot 2024-06-27 at 15 47 38" src="https://github.com/google/model-viewer/assets/2693840/c1fb2b13-d563-4b0c-8cfc-05ee3755b4aa"> | 
| ACES | <img width="785" alt="Screenshot 2024-06-27 at 15 48 55" src="https://github.com/google/model-viewer/assets/2693840/203bc006-1d76-4e5f-82ab-e71e640d9e06"> |

The asset: https://cloud.needle.tools/api/v1/public/8ab475a/8ab475a/caterpillar_work_boot.metal.glb (from https://sketchfab.com/3d-models/caterpillar-work-boot-d551ce74dcd24528a05cbb0f4b7434d7 with converted pbrSpecularGlossiness in gltf-transform)

### Version
latest

**Comment by elalish** [comment]
Indeed, this was an intentional change and was included both to get increased color accuracy and to address the feedback that models were looking "washed out". The key was realizing that Linear tone mapping does not actually preserve baseColors: PBR itself desaturates them by 4% on average. A correction for this was added to PBR Neutral so that baseColors can come through to the final render as unaltered as possible - find more details [here](https://modelviewer.dev/examples/tone-mapping#commerce). 

Of course, if you're taking Linear as "correct" (perhaps because the model was designed while looking at Linear output), then PBR Neutral will look dark. However, if you choose your sRGB base colors based on how the product should look on screen, then PBR Neutral will look correct and Linear will appear washed out. 

A side effect is that PBR Neutral now has a contrast toe like most Filmic tone mappers (including ACES), which gives deeper blacks and makes it easier to transition ACES-designed models to PBR Neutral without updating the GLB (which is pretty common). You'll notice with Linear that you can't get a matte, dielectric black darker than about `[30, 30, 30]` without adding AO or reducing lighting.

**Comment by hybridherbst** [comment]
Thanks a lot for the additional insights!

Doing some more tests, I think I can see what you mean with Linear being washed out when compared strictly to the 
 base color map, or when using the Furnace HDR., But I don't think Neutral comes closer to sRGB base colors than Linear under actual lighting conditions (right column):

At least to my eyes, Neutral looks oversatured and darkened compared to the base color reference. Am I missing something?

| Rendering Type | Reference | Furnace HDR | model-viewer Neutral HDR | 
| - | - | - | - |
| Base Color Map * Base Color Factor | ![image](https://github.com/google/model-viewer/assets/2693840/64b57bb3-8c55-436f-bb3f-26274468bcdb) |
| 3D: Base Color Map * Base Color Factor, Unlit, Linear | ![image](https://github.com/google/model-viewer/assets/2693840/bdad3c9c-e341-4a01-93c3-aecdb8a609f6) | 
| 3D: PBR Linear | | ![image](https://github.com/google/model-viewer/assets/2693840/1ca4c3f0-884d-42f7-89d7-31d06556d037) | ![image](https://github.com/google/model-viewer/assets/2693840/06994781-1aa6-4fbf-aebc-10bb3a41c673) |
| 3D: PBR Neutral | |  ![image](https://github.com/google/model-viewer/assets/2693840/24c4e4af-a184-4b3a-8d52-f9064ca69268) |  ![image](https://github.com/google/model-viewer/assets/2693840/90047a18-3d91-40f5-984a-9ae3bc66e778) |
| 3D: PBR AgX | | ![image](https://github.com/google/model-viewer/assets/2693840/886a5885-71b1-4c87-abc5-618ad778fab5) | ![image](https://github.com/google/model-viewer/assets/2693840/bbefeab1-d3b6-45b5-a4de-44503dc37e92) |

A side note, I would argue that 
> You'll notice with Linear that you can't get a matte, dielectric black darker than about [30, 30, 30]

is very much by design for the entire concept of PBR rendering – a shadow cast onto a black matte object needs to be visibly darker than the object, same as in the "real world".

**Reply by elalish** [reply]
>Doing some more tests, I think I can see what you mean with Linear being washed out when compared strictly to the
base color map, or when using the Furnace HDR., But I don't think Neutral comes closer to sRGB base colors than Linear under actual lighting conditions (right column):

I don't follow from looking at your images. You're saying that PBR Linear looks more like the reference than PBR Neutral? That's not what I see at all in your images above. Perhaps you can do a side-by-side just of the ones you're talking about and highlight what parts look more alike and different to you?

>a shadow cast onto a black matte object needs to be visibly darker than the object, same as in the "real world".

Well, this is a matter of contrast. In fact in the real world often shadows are hard to see on black objects. The problem we ran into was that for unshadowed black objects (the screen on a phone, the tire on a car) we couldn't get a black enough black - everything looked too gray. And we had a fair bit of feedback that this increased contrast was part of what users preferred about ACES.

**Reply by hybridherbst** [reply]
> You're saying that PBR Linear looks more like the reference than PBR Neutral?

Yes, indeed.

Here's unlit vs. neutral vs. linear copied from above –
In my eyes, linear matches albedo map much better than "Neutral" under regular lighting conditions (not furnace). "Neutral" oversaturates and darkens here, to the point of this becoming a different product, it becomes much more "yellow".

| Unlit (Base Color Map * Factor) | Neutral | Linear |
| - | - | - |
| ![image](https://github.com/google/model-viewer/assets/2693840/162c8a86-109c-4f2a-b7ce-928abcb5f4e5) | ![image](https://github.com/google/model-viewer/assets/2693840/685e212f-2ad8-48d4-afe8-b183be7f6a57) | ![image](https://github.com/google/model-viewer/assets/2693840/548bea3f-e25c-4cb6-ad8c-90a92f78bd7b) |

Here's an image with artifical brightness=3 to see that the left and right have a closer match than the middle one:

![image](https://github.com/google/model-viewer/assets/2693840/7aa9b70a-0385-4176-8c76-c97c8d1eea4a)

And here's a histogram – note the stronger blue-green separation on the "Neutral" image:
![image](https://github.com/google/model-viewer/assets/2693840/59301713-2906-4c5e-b1d0-36954313fe7c)

**Reply by elalish** [reply]
Thanks, I do see what you mean, though I don't completely agree. Still, it's worth looking into. First, how exactly are you doing these comparisons? It's slightly tricky with `<model-viewer>` because ACES has a different effective exposure than Linear/Neutral do, noted [here](https://modelviewer.dev/examples/lightingandenv/#toneMapping):

>For an apples-to-apples comparison of ACES to Neutral with custom lighting, set the Neutral exposure to 1 and the ACES exposure to 0.77 to account for ACES being artificially bright. This compensation is automatic for our built-in lighting.

Since tone mapping is non-linear, brightness applied before and after are not equivalent. 

Actually, I found a more serious issue - this boot is using the approximate conversion from spec gloss to metal rough, so it includes a specular factor and a non-physical IoR, which means the base color isn't the right reference on its own. In fact, I would bet that IoR = 1000 is the main reason for the discrepancy - we adjust for the average PBR desaturation of 4% based on IoR = 1.5. With 1000, that desaturation is gone and our compensation is going too far. 

I kind of hate that automatic spec-gloss conversion - technically it's the right way if you're actually doing something colorful with the specular channel, but 99% of the time it's just gray, so simply ditching the specular factor entirely and not touching IoR makes a more sensible PBR model.

**Reply by hybridherbst** [reply]
The file in question indeed has relevant color information in the specular channel, this data is not hand-authored but the result of a PBR scanning pipeline. Here's only the specular channel:

<img width="717" alt="image" src="https://github.com/google/model-viewer/assets/2693840/a8c9040a-0e60-4ea9-8ec0-2fa9505b7bae">

I might not know enough about the spec-gloss conversion (this particular file was converted with gltf-transform); my understanding was that `KHR_materials_specular` is a 100% replacement for the deprecated `KHR_materials_pbrSpecularGlossiness` the file used originally.

The extension notes that IOR should be set to 0, not 1000: https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_specular/README.md#specular-glossiness-materials

But in practice that doesn't seem to make a difference between 0.0 and 1000.0, at 1.5 (just tested out of curiosity) Neutral tonemapping becomes even darker.

| IOR | Result with Linear | Result with Neutral |
| - | - | - |
| 1000.0 | <img width="784" alt="image" src="https://github.com/google/model-viewer/assets/2693840/17887d29-45d5-4cd8-a58a-97e07fbb1b22"> | <img width="777" alt="image" src="https://github.com/google/model-viewer/assets/2693840/d0fdf850-1d5f-40d0-972d-7714c9ef87e2"> |
| 0.0 | <img width="797" alt="image" src="https://github.com/google/model-viewer/assets/2693840/dd55f88e-90ee-4b68-b7fc-c7c7533a3fd2"> | <img width="793" alt="image" src="https://github.com/google/model-viewer/assets/2693840/b0ea09d0-5ab7-4eb3-ac18-472f074e871c"> |
| 1.5 | <img width="775" alt="image" src="https://github.com/google/model-viewer/assets/2693840/41d7248c-2bfa-4650-8250-0c8685ffd5a9"> | <img width="755" alt="image" src="https://github.com/google/model-viewer/assets/2693840/107dd368-ae2a-4dcd-999f-1f49d189b6ca"> |

**Reply by elalish** [reply]
My point here is that there are currently a bunch of variables. PBR Neutral is darker than Linear intentionally - Linear cannot be considered a reference. The baseColor texture can be, but only if things are not too complicated (by things like specular and ior extensions). Can you choose or create a similar model that is baseline metallic-roughness so we can be precise about the differences?

**Reply by hybridherbst** [reply]
I'll try to find/make a better model.
